File: sdk/java/dagger-codegen-maven-plugin/src/main/java/io/dagger/codegen/introspection/TypeRef.java
package io.dagger.codegen.introspection;

import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.ParameterizedTypeName;
import com.palantir.javapoet.TypeName;
import java.util.List;

public class TypeRef {

  private TypeKind kind;
  private String name;
  private TypeRef ofType;

  public TypeKind getKind() {
    return kind;
  }

  public void setKind(TypeKind kind) {
    this.kind = kind;
  }

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public TypeRef getOfType() {
    return ofType;
  }

  public void setOfType(TypeRef ofType) {
    this.ofType = ofType;
  }

  public boolean isOptional() {
    return kind != TypeKind.NON_NULL;
  }

  public boolean isScalar() {
    TypeRef ref = this;
    if (ref.kind == TypeKind.NON_NULL) {
      ref = ref.ofType;
    }
    return ref.kind == TypeKind.SCALAR || ref.kind == TypeKind.ENUM;
  }

  public boolean isObject() {
    TypeRef ref = this;
    if (ref.kind == TypeKind.NON_NULL) {
      ref = ref.ofType;
    }
    return ref.kind == TypeKind.OBJECT;
  }

  public boolean isInterface() {
    TypeRef ref = this;
    if (ref.kind == TypeKind.NON_NULL) {
      ref = ref.ofType;
    }
    return ref.kind == TypeKind.INTERFACE;
  }

  public boolean isObjectOrInterface() {
    return isObject() || isInterface();
  }

  public boolean isList() {
    TypeRef ref = this;
    if (ref.kind == TypeKind.NON_NULL) {
      ref = ref.ofType;
    }
    return ref.kind == TypeKind.LIST;
  }

  public boolean isListOfObject() {
    TypeRef ref = this;
    if (ref.kind == TypeKind.NON_NULL) {
      ref = ref.ofType;
    }
    if (ref.kind != TypeKind.LIST) {
      return false;
    }
    ref = ref.getOfType();
    if (ref.kind == TypeKind.NON_NULL) {
      ref = ref.ofType;
    }
    return ref.isObject() || ref.isInterface();
  }

  public TypeRef getListElementType() {
    if (!isList()) {
      throw new IllegalArgumentException("Type is not a list");
    }
    TypeRef ref = this;
    while (ref.kind == TypeKind.NON_NULL || ref.kind == TypeKind.LIST) {
      ref = ref.ofType;
    }
    return ref;
  }

  public TypeName formatOutput() {
    return formatType(false, null);
  }

  public TypeName formatInput() {
    return formatType(true, null);
  }

  /** Format as input type, using the given expectedType for ID scalar resolution. */
  public TypeName formatInput(String expectedType) {
    return formatType(true, expectedType);
  }

  private TypeName formatType(boolean isInput, String expectedType) {
    if ("Query".equals(getName())) {
      return ClassName.bestGuess("Client");
    }
    switch (getKind()) {
      case SCALAR -> {
        switch (getName()) {
          case "String" -> {
            return ClassName.get(String.class);
          }
          case "Boolean" -> {
            return ClassName.get(Boolean.class);
          }
          case "Int" -> {
            return ClassName.get(Integer.class);
          }
          case "ID" -> {
            // Unified ID scalar: resolve to expected type if present
            if (isInput && expectedType != null && !expectedType.isEmpty()) {
              return ClassName.bestGuess(expectedType);
            }
            // When used as output (e.g. id() field), return the ID type
            return ClassName.bestGuess("ID");
          }
          default -> {
            if (!isInput) {
              return ClassName.bestGuess(getName());
            }
            return Helpers.convertScalarToObject(getName(), expectedType);
          }
        }
      }
      case OBJECT, ENUM, INPUT_OBJECT, INTERFACE -> {
        return ClassName.bestGuess(getName());
      }
      case LIST -> {
        return ParameterizedTypeName.get(
            ClassName.get(List.class), getOfType().formatType(isInput, expectedType));
      }
      default -> {
        return getOfType().formatType(isInput, expectedType);
      }
    }
  }

  public String getTypeName() {
    TypeRef ref = this;
    if (ref.kind == TypeKind.NON_NULL) {
      ref = ofType;
    }
    return ref.getName();
  }
}

File: sdk/csharp/src/Dagger.SDK.CodeGen/Types/TypeRef.cs
using System;
using System.Text.Json.Serialization;

namespace Dagger.SDK.CodeGen.Types;

public class TypeRef
{
    [JsonPropertyName("kind")]
    public required TypeKind Kind { get; set; }

    [JsonPropertyName("name")]
    public string? Name { get; set; }

    [JsonPropertyName("ofType")]
    public TypeRef? OfType { get; set; }

    public bool IsLeaf()
    {
        var tr = this;

        if (Kind == TypeKind.NON_NULL && OfType != null)
        {
            tr = OfType;
        }

        if (tr.Kind == TypeKind.ENUM)
        {
            return true;
        }

        if (tr.Kind == TypeKind.SCALAR)
        {
            return true;
        }

        return false;
    }

    public bool IsList()
    {
        var tr = this;

        if (Kind == TypeKind.NON_NULL && OfType != null)
        {
            tr = OfType;
        }

        if (tr.Kind == TypeKind.LIST)
        {
            return true;
        }

        return false;
    }

    public bool IsEnum()
    {
        var tr = this;

        if (Kind == TypeKind.NON_NULL && OfType != null)
        {
            tr = OfType;
        }

        if (tr.Kind == TypeKind.ENUM)
        {
            return true;
        }

        return false;
    }

    public bool IsInputObject()
    {
        var tr = this;

        if (Kind == TypeKind.NON_NULL && OfType != null)
        {
            tr = OfType;
        }

        if (tr.Kind == TypeKind.INPUT_OBJECT)
        {
            return true;
        }

        return false;
    }

    public bool IsScalar()
    {
        var tr = this;

        if (Kind == TypeKind.NON_NULL && OfType != null)
        {
            tr = OfType;
        }

        if (tr.Kind == TypeKind.SCALAR)
        {
            return true;
        }

        return false;
    }

    public bool IsObject()
    {
        var tr = this;

        if (Kind == TypeKind.NON_NULL && OfType != null)
        {
            tr = OfType;
        }

        if (tr.Kind == TypeKind.OBJECT)
        {
            return true;
        }

        return false;
    }

    public bool IsInterface()
    {
        var tr = this;

        if (Kind == TypeKind.NON_NULL && OfType != null)
        {
            tr = OfType;
        }

        if (tr.Kind == TypeKind.INTERFACE)
        {
            return true;
        }

        return false;
    }

    public TypeRef GetType_()
    {
        var tr = this;

        if (Kind == TypeKind.NON_NULL && OfType != null)
        {
            tr = OfType;
        }

        return tr;
    }

    public string Describe_()
    {
        return $" [{Name} {Kind}{OfType?.Describe_()}]";
    }
}
